cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -0,0 +1,507 @@
+      } else {
+        processed_count += min(rolling_buf_size, s->page.num_input_values - processed_count);
+      }
+
+      next_valid_count = gpuUpdateValidityOffsetsAndRowIndicesFlat<true, level_t>(
+        processed_count, s, sb, def, t, has_no_nulls);
+    }
+    // if we wanted to split off the skip_rows/num_rows case into a separate kernel, we could skip
+    // this function call entirely since all it will ever generate is a mapping of (i -> i) for
+    // nz_idx.  gpuDecodeValues would be the only work that happens.
+    else {
+      processed_count += min(rolling_buf_size, s->page.num_input_values - processed_count);
+      next_valid_count = gpuUpdateValidityOffsetsAndRowIndicesFlat<false, level_t>(
+        processed_count, s, sb, nullptr, t, false);
+    }
+    __syncthreads();
+
+    // decode the values themselves
+    gpuDecodeValues(s, sb, valid_count, next_valid_count, t);
+    __syncthreads();
+
+    valid_count = next_valid_count;
+  }
+}
+
+template <typename level_t>
+__global__ void __launch_bounds__(decode_block_size) gpuDecodePageDataFixedDict(
+  PageInfo* pages, device_span<ColumnChunkDesc const> chunks, size_t min_row, size_t num_rows)
+{
+  __shared__ __align__(16) page_state_s state_g;
+  __shared__ __align__(16) page_state_buffers_s<rolling_buf_size,  // size of nz_idx buffer
+                                                rolling_buf_size,  // dictionary
+                                                1>                 // unused in this kernel
+    state_buffers;
+
+  page_state_s* const s = &state_g;
+  auto* const sb        = &state_buffers;
+  int const page_idx    = blockIdx.x;
+  int const t           = threadIdx.x;
+  PageInfo* pp          = &pages[page_idx];
+
+  if (!(BitAnd(pages[page_idx].kernel_mask, decode_kernel_mask::FIXED_WIDTH_DICT))) { return; }
+
+  // must come after the kernel mask check
+  [[maybe_unused]] null_count_back_copier _{s, t};
+
+  if (!setupLocalPageInfo(s,
+                          pp,
+                          chunks,
+                          min_row,
+                          num_rows,
+                          mask_filter{decode_kernel_mask::FIXED_WIDTH_DICT},
+                          page_processing_stage::DECODE)) {
+    return;
+  }
+
+  __shared__ rle_run<level_t> def_runs[rle_run_buffer_size];
+  rle_stream<level_t, decode_block_size, rolling_buf_size> def_decoder{def_runs};
+
+  __shared__ rle_run<uint32_t> dict_runs[rle_run_buffer_size];
+  rle_stream<uint32_t, decode_block_size, rolling_buf_size> dict_stream{dict_runs};
+
+  // if we have no work to do (eg, in a skip_rows/num_rows case) in this page.
+  if (s->num_rows == 0) { return; }
+
+  bool const nullable     = is_nullable(s);
+  bool const has_no_nulls = nullable && no_nulls(s);
+
+  // initialize the stream decoders (requires values computed in setupLocalPageInfo)
+  level_t* const def = reinterpret_cast<level_t*>(pp->lvl_decode_buf[level_type::DEFINITION]);
+  if (!has_no_nulls) {
```

Comment:
I know I brought up the no_nulls, but @vuule is right about inverting it.

`!has_no_nulls` is `!(nullable && no_nulls)` which is `!nulllable || !no_nulls`, so if nullable is `false` we'll initialize the decoder? Did I do that right? This would be better as `if (has_nulls)`

Reply:
indeed! please take a look at https://github.com/rapidsai/cudf/pull/15159/commits/f415107953df6552c6b09c8f4dfed8efe1b9d177